### PR TITLE
[explorer] bug: fix display mosaic amount issue

### PR DESCRIPTION
**Pull request by AnthonyLaw**
## What was the issue?
- Mosaic amount wrong display, when sending transaction attached mosaic's Id and namespace. [read here ](https://github.com/symbol/explorer/issues/1051)
- Explorer filtered by mosaic id and showing the amount, instead of sum the amount balance.

## What's the fix?
- removed unused method `networkCurrencyMosaicBuilder`.
- sum the amount of mosaic filtered by mosaic id.
- minor refactor on helper method.

### CODE REVIEW

File: src/helper.js [hunk 1]
```diff
@@ -537,41 +539,35 @@ class helper {
...
-			throw new Error('Mosaic id does not match network Currency.');
+			if (idHex === http.networkCurrency.mosaicId) {
+				mosaicsFieldObject.push({
+					...mosaics[0],
+					rawAmount: UInt64.fromUint(sumAmount),
+					mosaicId: mosaics[0].id.toHex(),
+					amount: this.formatMosaicAmountWithDivisibility(sumAmount, http.networkCurrency.divisibility),
+					mosaicAliasName: http.networkCurrency.namespaceName
+				});
+			} else {
```
Comment (Jaguar0625) [inline src/helper.js:557]: why do you need special casing for network currency? why doesn't normal resolution logic work?

Comment (AnthonyLaw) [reply src/helper.js:557]: I excluded this because network currency info doesn't need a query from the network, so we can reduce 1 request for that (better perf).

In the `else` case, we requested mosaic info from the network.

Comment (Jaguar0625) [reply src/helper.js:557]: that makes sense, but why `sumAmount` in `if` case but `mosaics[0].amount` in `else case?

Comment (AnthonyLaw) [reply src/helper.js]: Good catch, it should in both.
I did some refactoring, and an extra unit test, to ensure those are covered.

File: src/helper.js [hunk 2]
```diff
@@ -537,41 +542,37 @@ class helper {
...
-			rawAmount: mosaic.amount,
-			mosaicId: mosaic.id.toHex(),
-			amount: this.formatMosaicAmountWithDivisibility(mosaic.amount, http.networkCurrency.divisibility),
-			mosaicAliasName: http.networkCurrency.namespaceName
-		};
+				mosaicsFieldObject.push({
+					...mosaicField,
+					amount: helper.formatMosaicAmountWithDivisibility(sumAmount, divisibility),
+					mosaicAliasName: MosaicService.extractMosaicNamespace({ mosaicId: mosaics[0].id.toHex() }, mosaicNames)
+				});
```
Comment (Jaguar0625) [inline src/helper.js:571]: you can go one step further refactoring. only remaining difference is divisibility used and derivation of mosaicAliasName. you can store those in local variables and then have only single push after if/else

File: __tests__/helper.spec.js [hunk 3]
```diff
@@ -1,18 +1,167 @@
...
+			// Act:
+			const result = await Helper.resolveMosaicId(mockNamespaceId);
+
+			// Assert:
+			expect(result).toStrictEqual(new MosaicId(http.networkCurrency.mosaicId).id);
+		});
+	});
+
+	describe('getNetworkCurrencyBalance should', () => {
+		it('returns network currency balance', async () => {
```
Comment (Jaguar0625) [inline __tests__/helper.spec.js:131]: add test where UNAVAILABLE is returned

File: __tests__/helper.spec.js [hunk 4]
```diff
@@ -1,18 +1,167 @@
...
+					amount: '6',
+					rawAmount: UInt64.fromUint(6),
+					mosaicId: mockTestMosaic.idHex,
+					mosaicAliasName: [mockTestMosaic.namespaceName]
+				}
+			]);
+		});
+	});
+
+	describe('resolveMosaicId should', () => {
```
Comment (Jaguar0625) [inline __tests__/helper.spec.js:106]: add test when mosaic being looked up is network currency mosaic

if possible, check if `NamespaceService.getLinkedMosaicId` mocked function is called or not

File: __tests__/helper.spec.js [hunk 5]
```diff
@@ -1,18 +1,167 @@
...
+import { stub, restore } from 'sinon';
+import { Mosaic, MosaicId, UInt64, NamespaceId, NamespaceName } from 'symbol-sdk';
 
-describe('Helper', () => {
+const mockTestMosaic = {
+	idHex: '6B2E9EAF2632AEC8',
+	namespaceName: 'namespace_mosaic'
+};
+
+describe.only('Helper', () => {
```
Comment (Jaguar0625) [inline __tests__/helper.spec.js:13]: drop .only

File: __tests__/helper.spec.js [hunk 6]
```diff
@@ -1,18 +1,167 @@
...
+			// Act:
 			const rgb = Helper.hslToRgb(hue, saturation, lightness);
 
-			//Assert:
+			// Assert:
 			expect(rgb).toEqual({R: 215, G: 158, B: 250});
 		});
 	});
+
+	describe('mosaicsFieldObjectBuilder should', () => {
```
Comment (Jaguar0625) [inline __tests__/helper.spec.js:29]: in describe, drop "should"
missing test where mosaics is empty

File: src/helper.js [hunk 7]
```diff
@@ -537,41 +542,37 @@ class helper {
...
-						...resolvedMosaic,
-						rawAmount: resolvedMosaic.amount,
-						mosaicId: resolvedMosaic.id.toHex(),
-						amount: helper.formatMosaicAmountWithDivisibility(resolvedMosaic.amount, divisibility),
-						mosaicAliasName: MosaicService.extractMosaicNamespace({ mosaicId: resolvedMosaic.id.toHex() }, mosaicNames)
-					});
-				}
-			}
-		}
+		const uniqueMosaicIds = [...new Set(resolvedMosaics.map(mosaic => mosaic.id.toHex()))];
```
Comment (Jaguar0625) [inline src/helper.js:545]: in line 525, why do you need to create new MosaicId, what is returned type of resolvedMosaic?

```
			const mosaicId = new MosaicId(resolvedMosaic.toHex()).id;
```

Comment (Jaguar0625) [reply src/helper.js:545]: in line 530, why are you filtering out network mosaic id?

Comment (Jaguar0625) [reply src/helper.js:545]: if you don't filter it out, uniqueMosaicIds would just be `[...new Set(resolvedMosaicIds)]`

File: __tests__/helper.spec.js [hunk 8]
```diff
@@ -1,18 +1,167 @@
...
+					mosaicId: http.networkCurrency.mosaicId,
+					rawAmount: UInt64.fromUint(1)
+				},
+				{
+					id: new MosaicId(mockTestMosaic.idHex).id,
+					amount: '4',
+					rawAmount: UInt64.fromUint(4),
+					mosaicId: mockTestMosaic.idHex,
+					mosaicAliasName: [mockTestMosaic.namespaceName]
+				}
```
Comment (Jaguar0625) [inline __tests__/helper.spec.js:70]: for readability, please use same property order for asserts of same object type

File: __tests__/helper.spec.js [hunk 9]
```diff
@@ -1,18 +1,167 @@
...
+					mosaicAliasName: http.networkCurrency.namespaceName,
+					mosaicId: http.networkCurrency.mosaicId,
+					rawAmount: UInt64.fromUint(1)
+				},
+				{
+					id: new MosaicId(mockTestMosaic.idHex).id,
+					amount: '4',
+					rawAmount: UInt64.fromUint(4),
+					mosaicId: mockTestMosaic.idHex,
+					mosaicAliasName: [mockTestMosaic.namespaceName]
```
Comment (Jaguar0625) [inline __tests__/helper.spec.js:69]: why is this array in this object but not previous one ?

File: src/helper.js (hunk 7 above)
Comment (Jaguar0625) [reply src/helper.js:545]: try to add test where 0 == resolvedMosaicIds.length

File: src/helper.js [hunk 10]
```diff
@@ -537,41 +542,37 @@ class helper {
...
-	static networkCurrencyMosaicBuilder = mosaic => {
-		if (!(mosaic instanceof Mosaic))
-			throw new Error('It required Mosaic instance.');
+			const sumAmount = mosaics.reduce((acc, cur) => acc + Number(cur.amount.toString()), 0);
 
-		if (mosaic.id.toHex() !== http.networkCurrency.mosaicId)
-			throw new Error('Mosaic id does not match network Currency.');
+			const mosaicField = {
+				...mosaics[0],
+				rawAmount: UInt64.fromUint(sumAmount),
```
Comment (Jaguar0625) [inline src/helper.js:554]: if there is precision issue, it's too late to convert to UInt64 here, and you probably want to use BigInt for arithmetic

File: src/helper.js [hunk 11]
```diff
@@ -537,41 +542,37 @@ class helper {
...
+				Object.assign(mosaicField, {
+					amount: this.formatMosaicAmountWithDivisibility(sumAmount, divisibility),
+					mosaicAliasName: MosaicService.extractMosaicNamespace({ mosaicId: mosaics[0].id.toHex() }, mosaicNames)
+				});
+			}
+
+			mosaicsFieldObject.push(mosaicField);
+		});
+
+		return mosaicsFieldObject;
```
Comment (Jaguar0625) [inline src/helper.js:575]: i think you can use map instead of forEach?

File: src/helper.js [hunk 12]
```diff
@@ -245,14 +248,13 @@ class helper {
...
 	static getNetworkCurrencyBalance = mosaics => {
-		let mosaic = mosaics.find(mosaic =>
-			mosaic.id.toHex() === http.networkCurrency.mosaicId ||
+		const networkCurrencyMosaic = mosaics.filter(mosaic => mosaic.id.toHex() === http.networkCurrency.mosaicId ||
 			(mosaic.id instanceof NamespaceId &&
-				mosaic.id.toHex() === http.networkCurrency.namespaceId));
+			mosaic.id.toHex() === http.networkCurrency.namespaceId));
 
-		let balance = mosaic !== undefined ? this.toNetworkCurrency(mosaic.amount) : Constants.Message.UNAVAILABLE;
+		const totalNetworkCurrencyAmount = networkCurrencyMosaic.reduce((acc, cur) => acc + Number(cur.amount.toString()), 0);
```
[the remark this answers is not in the document]
Comment (AnthonyLaw) [reply src/helper.js:255]: amount type is UInt64

File: src/helper.js [hunk 13]
```diff
@@ -245,14 +248,13 @@ class helper {
 	 * @returns {string} network currency balance.
 	 */
 	static getNetworkCurrencyBalance = mosaics => {
-		let mosaic = mosaics.find(mosaic =>
-			mosaic.id.toHex() === http.networkCurrency.mosaicId ||
+		const networkCurrencyMosaic = mosaics.filter(mosaic => mosaic.id.toHex() === http.networkCurrency.mosaicId ||
```
[the remark this answers is not in the document]
Comment (AnthonyLaw) [reply src/helper.js:251]: in the transaction, users are able to attach `UnresolvedMosaicId` (namespace id OR mosaic id)

File: src/helper.js (hunk 11 above)
Comment (AnthonyLaw) [reply src/helper.js:575]: That makes sense, I can just return rather than build and push to array.

File: src/helper.js (hunk 7 above)
Comment (AnthonyLaw) [reply src/helper.js:545]: > in line 525, why do you need to create a new MosaicId, what is returned type of resolvedMosaic?

it returns as `Id`, I will do some refactoring on this.

> in line 530, why are you filtering out network mosaic id?

those filtered mosaic ids need to call the API request mosaic info and namespace name, excluding network mosaic id, because we already have the information.

File: src/helper.js (hunk 10 above)
Comment (AnthonyLaw) [reply src/helper.js:554]: alright, I will take a look at this. 

`amount` from SDK is returned as `UInt64`, `rawAmount` will need to be UInt64 as well.

so we should convert `Uint64` -> `bigInt` and do arithmetic? after that convert back to `Uint64`?

File: __tests__/helper.spec.js (hunk 9 above)
Comment (AnthonyLaw) [reply __tests__/helper.spec.js:69]: so mosaic can have more than one alias name, this is the reason I use it as Array.

the previous one is network currency, which is that is not possible to have more than one alias name.

but it looks strange, I should convert it to the array.

File: src/helper.js (hunk 10 above)
Comment (Jaguar0625) [reply src/helper.js:548]: to make my point clearer, here is an example:

when you're doing arithmetic on Uint64s in JS, you want to do something like this [or equivalent]
> BigInt(Number.MAX_SAFE_INTEGER) + 2n - BigInt(Number.MAX_SAFE_INTEGER)
> 2

but you're effectively doing arithmetic with Number not BigInt, which could lead to wrong result:
> BigInt(Number.MAX_SAFE_INTEGER + 2) - BigInt(Number.MAX_SAFE_INTEGER)
> 1

File: __tests__/helper.spec.js [hunk 14]
```diff
@@ -119,15 +119,15 @@ describe('Helper', () => {
 			]);
 		});
 
-		it('returns empty array when mosaics not exist', async () => {
+		it('returns empty array when mosaics does not exist', async () => {
```
Comment (Jaguar0625) [inline __tests__/helper.spec.js:122]: does => do (because mosaics is plural)

**Review by Jaguar0625 — APPROVED** [review]

**Comment by AnthonyLaw** [comment]
@Jaguar0625 I rebased it to top of dev

**Review by Jaguar0625 — APPROVED** [review]
